File: PyVhdlUtil/VhdlParser.py
#   (C) 2014 Florian Huemer
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>

import subprocess
import re
import sys, traceback
import os  


from .VhdlEntity import VhdlEntity

class VhdlParser:
	
	def __init__(self):
		self.path_to_parser = os.path.dirname(os.path.realpath(__file__))+"/vhdlparser"
		self.error_list = []
		
	def get_component_from_file(self, path, component_name):
		out = subprocess.Popen([self.path_to_parser, "--component", component_name, str(path)], stdout=subprocess.PIPE).communicate()[0].decode("utf-8")
		lines = out.splitlines()
		component = self.create_vhdl_entity(component_name, lines)
		
		return component
		
	def get_entity_from_file (self, path, entity_name):
		out = subprocess.Popen([self.path_to_parser, "--entity", entity_name, str(path)], stdout=subprocess.PIPE).communicate()[0].decode("utf-8")
		lines = out.splitlines()
		entity = self.create_vhdl_entity(entity_name, lines)
		
		return entity

	def get_code_hierarchy_from_file (self, filename):
		out = subprocess.Popen([self.path_to_parser, str(filename)], stdout=subprocess.PIPE).communicate()[0].decode("utf-8")
		lines = out.splitlines()
		start_error_section = 0
		
		for line in lines:
			start_error_section += 1
			if (line == "[errors]"):
				break;
		
		code_hierarchy = self.create_code_hierarachy(lines[1:start_error_section-1])
		self.error_list = self.create_error_list(lines[start_error_section:len(lines)])
	
		return code_hierarchy
	
	def get_error_list(self) :
		return self.error_list
	
	def create_error_list(self, lines):
		error_list = []
		
		if(lines == None):
			return error_list
		
		for line in lines:
			m = re.search('([0-9]+)[ \t]+(.+)', line)
			#print(m.group(1));
			#print(m.group(2));
			line_no = int(m.group(1));
			msg = m.group(2);
			
			error_list.append(ParserError(msg, line_no))
	
		return error_list
	
	
	def create_code_hierarachy(self, lines):
		if (lines == None):
			return [];
	
		top_level_elements = []
		stack = [top_level_elements]
		
		for line in lines:
			m = re.search('([0-9]+)[ \t]+([0-9]+)[ \t]+(.+)[ \t]+(.+)', line)
		
			if (m != None):
				level = int(m.group(1));
				line_no = int(m.group(2));
				entry_type = m.group(3);
				name = m.group(4);
				
				entry = CodeHieararchyEntry(name, entry_type, line_no)
				
				while (len(stack) > level+1):
					stack.pop()
				
				stack[-1].append(entry)
				stack.append(entry.childeren)
		
		return top_level_elements

	def create_vhdl_entity(self, entity_name, lines):
		entity = VhdlEntity(entity_name)
		is_port_section = False
		
		for line in lines:
			if (line == "[generic]"):
				is_port_section = False
			elif (line == "[port]"):
				is_port_section = True
			else:
				if (is_port_section == False):
					m = re.search('([^\t]+)[ \t]([^\t]+)($|[ \t]([^\t]+)$)', line)
					if (m != None):
						name = m.group(1)
						data_type = m.group(2)
						default_value = m.group(4)
						#print (name + "  " +data_type +"  ["+str(default_value)+"]")
						if(default_value==""):
							entity.add_generic(name, data_type);
						else:
							entity.add_generic(name, data_type, default_value);
					else:
						print("error")
				elif (is_port_section == True):
					m = re.search('([^\t]+)[ \t]([0-9]+)[ \t]([^\t]+)($|[ \t]([^\t]+)$)', line)
					if (m != None):
						name = m.group(1)
						mode = int(m.group(2))
						data_type = m.group(3)
						default_value = m.group(4)
						#print (name + "  " +data_type +"  ["+default_value+"]")
						if(default_value==""):
							entity.add_port(name, mode, data_type);
						else:
							entity.add_port(name, mode, data_type, default_value);
					else:
						print("error")
						
		return entity;
		
class CodeHieararchyEntry:
	def __init__(self, name="", entry_type="", line=0):
		self.name = name
		self.line = line
		self.entry_type = entry_type
		self.childeren = []
	
class ParserError:
	def __init__(self, text="", line=0):
		self.text = text
		self.line = line 
		

File: vhdlparser/vhdlparser.c
/*
   (C) 2015 Florian Huemer

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>
*/

#include "vhdlparser.h"
#include "flexbison/parser.tab.h"
#include "flexbison/scanner.h"
#include "ast.h"
#include <malloc.h>

vhdl_parser_result* vhdl_parser_result_init()
{
	vhdl_parser_result *result = malloc(sizeof(vhdl_parser_result));
	result->enitiy_count = 0;
	result->architecture_count = 0;
	result->instance_count = 0;
	result->package_body_count = 0;
	result->package_count = 0;
	result->component_count = 0;
	
	result->errors = agc_new();
	result->ast = agc_new();
	
	return result;
}

int vhdl_parser(FILE *infile, vhdl_parser_result * result, int mode)
{
	yyscan_t myscanner;
	
	if (result == NULL || infile == NULL) {
		return -1;
	}
	
	scanner_yyextra extra;
	extra.yytextposition = 0;
	extra.parser_result = result;
	
	yylex_init_extra(&extra, &myscanner);
	yyset_in(infile, myscanner);
	yyparse(myscanner, result);
	yylex_destroy(myscanner);
	
	return 0;
}


/*
entities can only be on the "top-level" of an AST. So it is sufficient to
non-recursively iterate over the ast.
*/
struct node_entity* find_entity(void *ast, char *name)
{
	void* iter;
	agc_iterate_begin(ast, iter) {
		struct ast_node* node = (struct ast_node*)iter;
		if (node->type == TYPE_ENTITY) {
			if (!strcmp(((struct node_entity*)iter)->name, name)) {
				return iter;
			}
		}
	} agc_iterate_end(ast, iter);
	return NULL;
}


/*

*/
struct node_component* find_component(void *ast, char *name)
{
	if (ast == NULL) {
		return NULL;
	}
	void* iter;
	agc_iterate_begin(ast, iter) {
		struct ast_node* node = (struct ast_node*)iter;
		switch (node->type) {
			case TYPE_ARCHITECTURE:
			case TYPE_FUNCTION_BODY:
			case TYPE_PROCEDURE_BODY: 
			case TYPE_PROCESS:
			case TYPE_BLOCK:
			case TYPE_IF_GENERATE:
			case TYPE_FOR_GENERATE:
			case TYPE_PACKAGE_DECLARATION:
			case TYPE_PACKAGE_BODY:
			{
				struct node_component* c;
				c = find_component(((struct node_with_declaration_section*)iter)->declaration_section, name);
				if (c != NULL) {
					return c;
				}
			}
			break;
			case TYPE_COMPONENT:
				if (!strcmp(((struct node_component*)iter)->name, name)) {
					return iter;
				}
				break;
			
			default:
				break;
		}
	} agc_iterate_end(ast, iter);
	return NULL;
}






